sdk/ai/Azure.AI.Inference/tests/Utilities/ValidatingActivityListener.cs

```diff
@@ -0,0 +1,239 @@
+        {
+            m_activityListener = new ActivityListener()
+            {
+                ActivityStopped = ActivityStopped,
+                ShouldListenTo = s => s.Name == OpenTelemetryConstants.ClientName,
+                Sample = (ref ActivityCreationOptions<ActivityContext> _) => ActivitySamplingResult.AllDataAndRecorded
+            };
+            ActivitySource.AddActivityListener(m_activityListener);
+        }
+
+        public void Dispose()
+        {
+            m_activityListener.Dispose();
+        }
+
+        public void ValidateStartActivity(ChatCompletionsOptions requestOptions, Uri endpoint, bool traceEvents)
+        {
+            Activity activity = m_listeners.Single();
+            Assert.NotNull(activity);
+            // Validate all tags
+            ValidateTag(activity, GenAiSystemKey, GenAiSystemValue);
+            ValidateTag(activity, GenAiResponseModelKey, requestOptions.Model);
+            ValidateTag(activity, ServerAddressKey, endpoint.Host);
+            if (endpoint.Port != 443)
+                ValidateTag(activity, ServerPortKey, endpoint.Port);
+            else
+                Assert.IsNull(activity.GetTagItem(ServerPortKey));
+            ValidateTag(activity, GenAiOperationNameKey, "chat");
+            ValidateTag(activity, GenAiRequestMaxTokensKey, requestOptions.MaxTokens);
+            ValidateFloatTag(activity, GenAiRequestTemperatureKey, requestOptions.Temperature);
+            if (requestOptions.AdditionalProperties.TryGetValue("top_p", out BinaryData topP))
+            {
+                ValidateFloatTag(activity, GenAiRequestTopPKey, JsonSerializer.Deserialize<float?>(topP));
+            }
+            // Validate events
+            if (traceEvents)
+                ValidateChatMessageEvents(activity, requestOptions.Messages);
+            else
+                foreach (ChatRequestMessage message in requestOptions.Messages)
+                {
+                    ValidateNoEventsWithName($"gen_ai.{message.Role}.message");
+                }
+        }
+
+        /// <summary>
+        /// Remove junk added adding string as a tag.
+        /// </summary>
+        /// <param name="value">The string to be cleaned.</param>
+        /// <returns></returns>
+        private static string CleanString(string value)
+        {
+            return value.Replace("\\u0022", "\"").Replace("\\\\n", "\\n").Replace("\\\\u", "\\u").Trim('\"');
+        }
+
+        public void ValidateResponseEvents(AbstractRecordedResponse response, bool traceEvents)
+        {
+            Activity activity = m_listeners.Single();
+            if (response.IsEmpty)
+            {
+                ValidateNoEventsWithName(GenAiChoice);
+                return;
+            }
+            ValidateTag(activity, GenAiResponseIdKey, response.Id);
+            ValidateTag(activity, GenAiResponseModelKey, response.Model);
+            ValidateTag(activity, GenAiResponseFinishReasonKey, response.FinishReason);
```

Comment:
```suggestion
            ValidateTag(activity, GenAiResponseFinishReasonsKey, response.FinishReason);
```

it should be an array

Reply:
Fixed.